src/main/java/org/truffleruby/core/array/ArrayNodes.java

```diff
@@ -551,22 +551,35 @@ public DynamicObject clear(DynamicObject array,
     @ImportStatic(ArrayGuards.class)
     public abstract static class CompactNode extends ArrayCoreMethodNode {
 
-        @Specialization(guards = { "strategy.matches(array)", "!strategy.accepts(nil())" }, limit = "ARRAY_STRATEGIES")
+        @Specialization(guards = { "strategy.matches(array)", "strategy.isPrimitive()" }, limit = "STORAGE_STRATEGIES")
         public DynamicObject compactPrimitive(DynamicObject array,
                 @Cached("of(array)") ArrayStrategy strategy) {
             final int size = strategy.getSize(array);
             Object store = strategy.newMirror(array).extractRange(0, size).getArray();
+            strategy.setStore(array, strategy.newMirror(array).extractRange(0, size).getArray());
             return createArray(store, size);
         }
 
-        @Specialization(guards = { "strategy.matches(array)", "strategy.accepts(nil())" }, limit = "ARRAY_STRATEGIES")
+        @Specialization(guards = { "strategy.isStorageMutable()", "strategy.matches(array)", "!strategy.isPrimitive()" }, limit = "STORAGE_STRATEGIES")
         public Object compactObjects(DynamicObject array,
                 @Cached("of(array)") ArrayStrategy strategy) {
-            // TODO CS 9-Feb-15 by removing nil we could make this array suitable for a primitive array storage class
+            return compactElement(array, strategy, strategy);
+        }
+
+        @Specialization(guards = { "!strategy.isStorageMutable()", "strategy.matches(array)", "!strategy.isPrimitive()" }, limit = "STORAGE_STRATEGIES")
```

Comment:
Could you move the isStorageMutable() at the end so we have them in the same order in each specialization and it's easier to see they are complementary? (But see below to remove the difference)